### DISCUSSION TITLE
Choosing source IP address of outgoing packets on machine with multiple IPs.

### DISCUSSION THREAD

**Discussion by davidecorradini** [mitmproxy/mitmproxy#5364] [Q&A]
Hello,
I am using mitmproxy on an Ubuntu machine with two IP addresses on the same interface (e.g., 192.168.0.10/24 and 192.168.0.11/24). How can I choose the source IP address of outgoing packets?
I tried with the argument `--listen-host`, but it does not change the source IP address.
Can you please help me with this?
Thank you in advance.
Regards.

**Comment by mhils** [comment]
At the moment mitmproxy just picks the default gateway, although I'd certainly be happy to accept PRs that improve the situation. :)

**Reply by meitinger** [reply]
How about allowing `Server.sockname` to be prepopulated. This will allow `command.OpenConnection` to specify it, `ServerConnectHook` to override it, and then we pass it on in https://github.com/mitmproxy/mitmproxy/blob/4a2ca7563108d2256909309c770feb0c5b481397/mitmproxy/proxy/server.py#L175 with `local_addr`. We could also introduce a `--connect-addr` option, where `Proxyserver` will populate `sockname` in https://github.com/mitmproxy/mitmproxy/blob/4a2ca7563108d2256909309c770feb0c5b481397/mitmproxy/addons/proxyserver.py#L387 if the option is set and the field is still `None`.

If that sounds good, I'll open a PR (and also implement `local_addr` for UDP's `open_connection`).

**Reply by mhils** [reply]
I absolutely agree on the first part of making this available via the API, but let's not add an additional option for now. I don't think we even need to prepopulate it, a default of `None` should be fine to just select the default, no? :)

**Reply by meitinger** [reply]
If we don't add a command line option, you're right, there is no need for `Proxyserver` to set anything. (`None` would also be default if neither `connect-addr` is specified nor a plugin sets `sockname` in `server_connect`.)

The reason I suggested the command line option was that not everyone might be able to code a plugin, so having an option would enable them to make use of the functionality without writing code.

**Reply by mhils** [reply]
Ok - "without writing code" is a somewhat reasonable argument. Let's add an option as well if that helps. :)

**Comment by meitinger** [comment]
PR #5366 created. I'm a bit unsure about the name `connect_addr` for the command line option, but since we have `listen-host` and `listen-port`, 'connect' would be the counterpart.
